standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+| Maximum total rate limit of concurrent active memberships | `R_{max}`  | TBD     | messages per `epoch`              | Yes         | Can be replaced with `N_{max}` |
+| Maximum number of concurrent active memberships           | `N_{max}`  | `10000` | units                             | Yes         | Can be replaced with `R_{max}` |
+| Rate limit for low-tier                                   | `R_{low}`  | `20`    | messages per `epoch`              | TBD         |                                |
+| Rate limit for mid-tier                                   | `R_{mid}`  | `200`   | messages per `epoch`              | TBD         |                                |
+| Rate limit for high-tier                                  | `R_{high}` | `600`   | messages per `epoch`              | TBD         |                                |
+| Unit price                                                | `p_u`      | TBD     | `USD` per `1` message per `epoch` | Yes         |                                |
+| Membership expiration term                                | `T`        | `90`    | days                              | Yes         |                                |
+| Grace period                                              | `G`        | `30`    | days                              | Yes         |                                |
+
+### Discussion on some parameter values
+
+#### Epoch length
+
+Epoch length is a global parameter set in the smart contract.
+Rate limits are defined in terms of the maximum allowed messages per epoch.
+There is a trade-off between short and long epochs.
+
+On the one hand, longer epochs allow for short-term spikes,
+which allows for more flexibility.
+Arguably, short-term bursts followed by a period of silence is a common pattern for some use cases.
+Usage peaks are expected to average out over longer time periods,
+allowing us to reason about network utilization on a longer time scale.
+
+On the other hand, long epochs lead to more memory requirements for nodes due to the growth of the nullifier log.
+Each message contains a nullifier that proves its validity in terms of RLN.
+Within the latest epoch, each relaying node stores all nullifiers to detect users exceeding the rate limit.
+Each nullifier plus metadata is `128` bytes (per message).
+This means that one user with a `1` message per second rate limit (high-tier) generates up to `600 * 128 / 1024 = 75 KiB` of nullifier log data per 10-min epoch.
+
+#### Maximum total message limit / concurrently active memberships
+
+We may want to limit the total rate limit available to all users.
+The rationale is that the total network bandwidth is a limited resource.
+The total active rate limit should not exceed the network's real capabilities.
+
+There may be two ways to implement a global cap:
+- limit the total rate limits;
```

Comment:
Is that possible?

Reply:
It might also be worth clarifying how such total rate limits would be calculated - e.g. max number of memberships * msg size * msg/s = XYZ kB/s - since it seems like this is based on max bandwith of the network, but that depends on the max msg size, which is defined elsewhere

Reply:
> Is that possible?

Why wouldn't it be possible? Say, the contract would maintain a variable that is incremented at every membership registraction and decremented at every ~~expiration~~ invalidation (I mean, expiration is insufficient because of grace period). If a new membership registration would push the variable over the hard-coded limit (modifiable by the _Owner_), it would fail.

Reply:
> what we think is reasonable now, i.e. limit the total number of active membership

Perhaps I haven't phrased this well enough: my question here is which of the parameters, or both, we want to limit: the total number of active memberships, or their total rate limit. I'm not suggesting anything like

> adapting membership availability based on current total rate limit

Rather, I simply hesitate between two possible types of restrictions:

1. the total number of active memberships `n` cannot exceed `N`;
2. the total rate limit of active memberships `x` cannot exceed `X` messages per epoch,

where `N` and `X` are hard-coded, _Owner_-modifiable variables.

The former can be somewhat imprecise, as `N` high-tier memberships are different than `N` low-tier memberships in the total burden they impose on the network. That said, there are some aspects (e.g., tree size?) that increase proportionally to `n`, not to `x`. Therefore, I'd suggest using both limits, and fail a registration attempt if a new membership would violate either one.

Reply:
Ah yes thanks. Not sure how I interpreted it at first read but makes sense now.

Yes I agree that ultimately, we want to limit the total rate, ie, the sum of allowed rate limit of all active membership.

Maybe it's a question of whether we want to have this now or later. I think if it's not too complicated, we do it now.

If we limit the total rate, so not sure if it's helpful to also limit the total number of members. Except if it's technically necessary. Eg at some point we discussed to initialize the merkle tree at deployement with zerod leaf. In this case we need to set a size at deployment.

In terms of decrementing, I assume it would happen in 2 scenarios:

- current member withdraw their deposit
- OR, new member takes an expired slot with a different rate limit

Reply:
> if it's helpful to also limit the total number of members. Except if it's technically necessary.

My concern is that if we don't limit the total number of members, we'd enable some DoS attack where the tree is polluted by a large number of memberships. Although we can likely prevent this by introducing a minimal rate limit (assuming a curve-based scheme). If the total rate limit must be below `X`, and each membership must have a rate limit above `Y`, then the total number of memberships cannot be higher than `X/Y`.

In short: I would suggest implementing an upper bound on the total rate limit of all^ memberships and a lower bound on the rate limit of one membership.

^ All, not just _active_: memberships in _grace-period_ and _expired_ states also allow sending messages, therefore, I think they should count towards the limit as well. _Erased_ memberships don't count (they've been removed from the tree at that point).